src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -30,21 +34,47 @@ private sealed class State
             // set immediately in ctor, but we need a GCHandle to State in order to create the handle.
             public SafeMsQuicListenerHandle Handle = null!;
 
-            public readonly SafeMsQuicConfigurationHandle ConnectionConfiguration;
+            public readonly SafeMsQuicConfigurationHandle? ConnectionConfiguration;
             public readonly Channel<MsQuicConnection> AcceptConnectionQueue;
 
-            public bool RemoteCertificateRequired;
-            public X509RevocationMode RevocationMode = X509RevocationMode.Offline;
-            public RemoteCertificateValidationCallback? RemoteCertificateValidationCallback;
+            public QuicOptions ConnectionOptions = new QuicOptions();
+            public SslServerAuthenticationOptions AuthenticationOptions = new SslServerAuthenticationOptions();
+            public QuicListener.ServerOptionsSelectionCallback? ServerOptionsSelectionCallback;
+            public object? ServerOptionsSelectionCallbackState;
 
             public State(QuicListenerOptions options)
             {
-                ConnectionConfiguration = SafeMsQuicConfigurationHandle.Create(options);
-                if (options.ServerAuthenticationOptions != null)
+                ConnectionOptions.IdleTimeout = options.IdleTimeout;
+                ConnectionOptions.MaxBidirectionalStreams = options.MaxBidirectionalStreams;
+                ConnectionOptions.MaxUnidirectionalStreams = options.MaxUnidirectionalStreams;
+                ServerOptionsSelectionCallback = options.ServerOptionsSelectionCallback;
+                ServerOptionsSelectionCallbackState = options.ServerOptionsSelectionCallbackState;
+
+                bool delayConfiguration = false;
+
+                if (options.ServerOptionsSelectionCallback != null)
                 {
-                    RemoteCertificateRequired = options.ServerAuthenticationOptions.ClientCertificateRequired;
-                    RevocationMode = options.ServerAuthenticationOptions.CertificateRevocationCheckMode;
-                    RemoteCertificateValidationCallback = options.ServerAuthenticationOptions.RemoteCertificateValidationCallback;
+                    delayConfiguration = true;
```

Comment:
Is there a good reason to not just always delay configuration? Does it matter in practice? Seems like that would simplify the logic.

Reply:
When certificate is provided upfront we need to do this only once and we re-se same connectionConfiguration for all incoming connection. It does not need to be here but if there is something wrong with provided option I feel it is better to throw sooner than later.